tachyon/math/matrix/prime_field_num_traits.h

```diff
@@ -1,30 +1,67 @@
 #ifndef TACHYON_MATH_MATRIX_PRIME_FIELD_NUM_TRAITS_H_
 #define TACHYON_MATH_MATRIX_PRIME_FIELD_NUM_TRAITS_H_
 
+#include <type_traits>
+
 #include "third_party/eigen3/Eigen/Core"
 
 #include "tachyon/math/finite_fields/finite_field_forwards.h"
 
 namespace Eigen {
 
+template <typename T, typename SFINAE = void>
+struct CostCalculator;
+
 template <typename Config>
-struct NumTraits<tachyon::math::PrimeField<Config>>
-    : GenericNumTraits<tachyon::math::PrimeField<Config>> {
+struct CostCalculator<tachyon::math::PrimeField<Config>,
+                      std::enable_if_t<Config::kModulusBits <= 32>> {
+  constexpr static int ComputeReadCost() {
+    return NumTraits<uint32_t>::ReadCost;
+  }
+  constexpr static int ComputeAddCost() {
+    // In general, c = (a + b) % M = (a + b) > M ? (a + b) - M : (a + b)
+    return NumTraits<uint32_t>::AddCost * 3 / 2;
+  }
+  constexpr static int ComputeMulCost() {
+    // In general, c = (a * b) % M = (a * b) - [(a * b) / M] * M
+    return 4 * NumTraits<uint32_t>::MulCost + NumTraits<uint32_t>::AddCost;
+  }
+};
+
+template <typename Config>
+struct CostCalculator<tachyon::math::PrimeField<Config>,
+                      std::enable_if_t<(Config::kModulusBits > 32)>> {
   constexpr static size_t kLimbNums =
       tachyon::math::PrimeField<Config>::kLimbNums;
```

Comment:
Why not create just one `CostCalculator` and set `kLimbNums` to 1 & set `using NumTraitsType = std::conditional_t <Config::kModulusBits <= 32, NumTraits<uint32_t>, NumTraits<uint64_t>>` if `Config::kModulusBits <= 32`?

Reply:
This is because the types instantiated with `NumTraits` are different. One is `uint64_t` the other is `uint32_t`.